prog/vm/github_runner.rb

```diff
@@ -0,0 +1,162 @@
+        installation_id: installation.id,
+        repository_name: repository_name,
+        label: label,
+        vm_id: vm_st.id
+      ) { _1.id = ubid.to_uuid }
+
+      Strand.create(prog: "Vm::GithubRunner", label: "start") { _1.id = github_runner.id }
+    end
+  end
+
+  def vm
+    @vm ||= github_runner.vm
+  end
+
+  def github_client
+    @github_client ||= Github.installation_client(github_runner.installation.installation_id)
+  end
+
+  def before_run
+    when_destroy_set? do
+      if strand.label != "destroy"
+        hop_destroy
+      end
+    end
+  end
+
+  label def start
+    register_deadline(:wait, 10 * 60)
+
+    hop_wait_vm
+  end
+
+  label def wait_vm
+    nap 5 unless vm.strand.label == "wait"
+    vm.sshable.update(host: vm.ephemeral_net4 || vm.ephemeral_net6&.nth(2).to_s)
+    hop_bootstrap_rhizome
+  end
+
+  label def bootstrap_rhizome
+    bud Prog::BootstrapRhizome, {"target_folder" => "common", "subject_id" => vm.id, "user" => "runner"}
+    hop_wait_bootstrap_rhizome
+  end
+
+  label def wait_bootstrap_rhizome
+    reap
+    hop_install_actions_runner if leaf?
+    donate
+  end
+
+  # TODO: Move this to golden image too
+  label def install_actions_runner
+    vm.sshable.cmd("curl -o actions-runner-linux-x64-2.308.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.308.0/actions-runner-linux-x64-2.308.0.tar.gz")
+    vm.sshable.cmd("echo '9f994158d49c5af39f57a65bf1438cbae4968aec1e4fec132dd7992ad57c74fa  actions-runner-linux-x64-2.308.0.tar.gz' | shasum -a 256 -c")
+    vm.sshable.cmd("tar xzf ./actions-runner-linux-x64-2.308.0.tar.gz")
+
+    hop_register_runner
+  end
+
+  label def register_runner
+    unless github_runner.runner_id
+      # We use generate-jitconfig instead of registration-token because it's
+      # recommended by GitHub for security reasons.
+      # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners
+      data = {name: github_runner.ubid.to_s, labels: ["ubicloud"], runner_group_id: 1}
+      response = github_client.post("/repos/#{github_runner.repository_name}/actions/runners/generate-jitconfig", data)
+      github_runner.update(runner_id: response[:runner][:id], ready_at: Time.now)
+      vm.sshable.cmd("common/bin/daemonizer 'sudo -u runner /home/runner/run.sh --jitconfig #{response[:encoded_jit_config].shellescape}' runner-script")
+    end
+
+    case vm.sshable.cmd("common/bin/daemonizer --check runner-script")
+    when "Succeeded", "InProgress"
+      hop_wait
+    when "NotStarted", "Failed"
+      github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{github_runner.runner_id}")
+      github_runner.update(runner_id: nil, ready_at: nil)
+    end
+    nap 10
+  end
+
+  label def wait
+    # If the runner doesn't pick a job in two minutes, destroy it
+    if github_runner.job_id.nil? && Time.now > github_runner.ready_at + 60 * 2
```

Comment:
I meant can we ask GitHub, "hey do you have any job waiting for a runner"? If it says yes, we wouldn't delete the runner because having a job waiting runner and having a runner waiting for a job at the same time probably means that GitHub already send us a request and it failed for some reason and it will send it again in 60 seconds.

Reply:
Unfortunately GitHub doesn't have such an endpoint.

Some possible solutions:

### Solution 1:
`GET /repos/{owner}/{repo}/actions/runs?status=queued` returns list of queued runs. But run doesn't have label info to determine jobs for `ubicloud` runners. We need individual job data.

We can get jobs for each run from previous endpoint using `GET /repos/{owner}/{repo}/actions/runs/{run_id}/jobs`. It has individual job info with label data. But we need to send this request for each run.

Also these paths require `/{owner}/{repo}/`, so we need to determine repository. We can get list of repository for each GitHub installation using `GET /installation/repositories`

But we need to send INSTALLATION_COUNT + REPOSITORY_COUNT + QUEUED_RUN_COUNT request to find queued and not started jobs. It looks too much to me.

### Solution 2
We are able get list of webhook deliveries using `GET /app/hook/deliveries` endpoint.
We can try to use list of deliveries to determine queued but not started job. But if it's in that list and our app isn't down, we get that webhook event too.

### Solution 3
We can create another table such as `GithubWorkflowJob` to keep jobs. When we receive `workflow_job.queued` event, we ca create a record. We can check its status until completion, and create extra runners if `GithubWorkflowJob.count > GithubRunner.count`

Reply:
I agree it is too much work to handle this case. Let's keep it in current way and we can think a solution if we hit this too much.

Can we log something here so that we can know frequency of such events?

Reply:
I added
```
puts "Destroying GithubRunner[#{github_runner.ubid}] because it does not pick a job in two minutes"
```